testing/src/ops_main_mock.py

```diff
@@ -0,0 +1,288 @@
+        event=event,
+        context=context,
+        charm_spec=charm_spec,
+    )
+    debug = "JUJU_DEBUG" in os.environ
+    setup_root_logging(model_backend, debug=debug)
+    # ops sets sys.excepthook to go to Juju's debug-log, but that's not useful
+    # in a testing context, so reset it.
+    sys.excepthook = sys.__excepthook__
+    ops_logger.debug(
+        "Operator Framework %s up and running.",
+        ops.__version__,
+    )  # type:ignore
+
+    metadata = (charm_dir / "metadata.yaml").read_text()
+    actions_meta = charm_dir / "actions.yaml"
+    if actions_meta.exists():
+        actions_metadata = actions_meta.read_text()
+    else:
+        actions_metadata = None
+
+    meta = CharmMeta.from_yaml(metadata, actions_metadata)
+
+    # ops >= 2.10
+    if inspect.signature(ops.model.Model).parameters.get("broken_relation_id"):
+        # If we are in a RelationBroken event, we want to know which relation is
+        # broken within the model, not only in the event's `.relation` attribute.
+        broken_relation_id = (
+            event.relation.id  # type: ignore
+            if event.name.endswith("_relation_broken")
+            else None
+        )
+
+        model = ops.model.Model(
+            meta,
+            model_backend,
+            broken_relation_id=broken_relation_id,
+        )
+    else:
+        ops_logger.warning(
+            "It looks like this charm is using an older `ops` version. "
+            "You may experience weirdness. Please update ops.",
+        )
+        model = ops.model.Model(meta, model_backend)
+
+    charm_state_path = charm_dir / CHARM_STATE_FILE
+
+    # TODO: add use_juju_for_storage support
+    store = ops.storage.SQLiteStorage(charm_state_path)
+    framework = ops.Framework(store, charm_dir, meta, model)
+    framework.set_breakpointhook()
+    return framework
+
+
+def setup_charm(
+    charm_class: Type[ops.CharmBase], framework: ops.Framework, dispatcher: _Dispatcher
+):
+    sig = inspect.signature(charm_class)
+    sig.bind(framework)  # signature check
+
+    charm = charm_class(framework)
+    dispatcher.ensure_event_links(charm)
+    return charm
+
+
+def setup(
+    state: "State",
+    event: "_Event",
+    context: "Context",
+    charm_spec: "_CharmSpec[ops.CharmBase]",
+):
+    """Setup dispatcher, framework and charm objects."""
+    charm_class = charm_spec.charm_type
+    charm_dir = _get_charm_dir()
+
+    try:
+        dispatcher = _Dispatcher(charm_dir)  # type: ignore
+    except TypeError:
+        # ops 2.16+
+        import ops.jujucontext  # type: ignore
+
+        dispatcher = _Dispatcher(
+            charm_dir,
+            ops.jujucontext._JujuContext.from_dict(os.environ),  # type: ignore
+        )
+    dispatcher.run_any_legacy_hook()
+
+    framework = setup_framework(charm_dir, state, event, context, charm_spec)
+    charm = setup_charm(charm_class, framework, dispatcher)
+    return dispatcher, framework, charm
+
+
+class Ops:
```

Comment:
perhaps we can do that later, but now is a good time to trim down ops_main_mock and start using directly some of the newer stuff in ops

Reply:
I think it is good to do that soon, but not in this PR, which is already very large/complex.